Invite panel

devmount opened the pull request:
<!--
* Filling out the template is required.
* All new code must have been tested to ensure against regressions
-->

## Description of the Change

This change introduces admin only views. Those are currently protected by a dependency checking, if the current users email address is whitelisted in a .env var `APP_ADMIN_ALLOW_LIST` (comma separated list of email addresses).

New route `/admin/invites`:
![image](https://github.com/thunderbird/appointment/assets/5441654/eb45ac6f-3c77-43f9-a776-08fd1bea5bdc)

Invite codes list is paginated, filterable, and codes can be bulk generated, copied and revoked.

New route `/admin/subscribers`:
![Untitled](https://github.com/thunderbird/appointment/assets/5441654/3c7a21ac-2f82-4ae6-8f98-4550d594487f)

For now, subscribers can only be listed. Showing invite status and disabling them is still todo.

## Benefits

Administration tools for quick management of invites and subscribers.

## Applicable Issues

Closes #362

MelissaAutumn commented:
I'll fix up the PR with any changes needed!

MelissaAutumn commented:
I accidentally rebased this on my old changes, so I had to go through and download / apply patch files from the now headless remote. It should be fixed now, if not a little out of order. 

Apologies!

MelissaAutumn commented:
Idea pivot! Have a subscriber panel with a invite input at the footer. This invite function will create the subscriber and email them about Appointment. 

I wrote a generic data table component that takes in structured data and displays it. There's some extra functionality I decided to cut for now, but are still implemented in the data table. I'm sure there's an easier way to do filtering than keeping the source list and a filtered version but I cannot think of a good way to do it right now. 

If someone could pretty-ify the placement of the data filters that would be great 😅 

<img width="1382" alt="image" src="https://github.com/thunderbird/appointment/assets/97147377/98779517-4ac0-41f1-973c-6d10cac41878">

MelissaAutumn commented on backend/src/appointment/routes/auth.py:
```diff
@@ -237,6 +241,13 @@ def me(
+@router.post("/permission-check")
+def permission_check(_: Subscriber = Depends(get_admin_subscriber)):
+    """Checks if they have admin permissions"""
+    return True
```
The check is done in get_admin_subscriber, and will error out. FastAPI has a weird relationship with using DI over middleware...

MelissaAutumn commented:
Additionally all admin checks will fail if `APP_ADMIN_ALLOW_LIST` is empty.

devmount commented:
No worries about the rebase, everything looks fine now. Sorry you had some additional effort because of that.

The new subscriber panel is nice, also simplifies things for not having different admin pages for now. I fixed up some alignment and set the horizontal overflow scrollbar to auto (as for some systems, the scrollbar would be always visible with "scroll" 🙈 ). Also updated the language files (my IDE sorted the JSON keys, sorry about that 😇 ).

So looking fine now! Idea for the data filters: We could integrate them into the table header beneath the column label. But I think it's also fine to have them above the table 🤷🏻

MelissaAutumn commented:
Thank you! Also thank you for re-sorting the language files as always!
